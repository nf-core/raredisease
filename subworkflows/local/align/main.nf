//
// Map to reference
//

include { FASTP                      } from '../../../modules/nf-core/fastp/main'
include { ALIGN_BWA_BWAMEM2_BWAMEME  } from '../align_bwa_bwamem2_bwameme'
include { ALIGN_SENTIEON             } from '../align_sentieon'
include { SAMTOOLS_VIEW              } from '../../../modules/nf-core/samtools/view/main'
include { ALIGN_MT                   } from '../align_MT'
include { ALIGN_MT as ALIGN_MT_SHIFT } from '../align_MT'
include { CONVERT_MT_BAM_TO_FASTQ    } from '../convert_mt_bam_to_fastq'

workflow ALIGN {
    take:
        ch_reads                 // channel: [optional] [ val(meta), [path(reads)]  ]
        ch_alignments            // channel: [optional] [ val(meta), [path(bam),path(bai)]  ]
        ch_genome_fasta          // channel: [mandatory] [ val(meta), path(fasta) ]
        ch_genome_fai            // channel: [mandatory] [ val(meta), path(fai) ]
        ch_genome_bwaindex       // channel: [mandatory] [ val(meta), path(index) ]
        ch_genome_bwamem2index   // channel: [mandatory] [ val(meta), path(index) ]
        ch_genome_bwamemeindex   // channel: [mandatory] [ val(meta), path(index) ]
        ch_genome_dictionary     // channel: [mandatory] [ val(meta), path(dict) ]
        ch_mt_bwaindex           // channel: [mandatory] [ val(meta), path(index) ]
        ch_mt_bwamem2index       // channel: [mandatory] [ val(meta), path(index) ]
        ch_mt_dictionary         // channel: [mandatory] [ val(meta), path(dict) ]
        ch_mt_fai                // channel: [mandatory] [ val(meta), path(fai) ]
        ch_mt_fasta              // channel: [mandatory] [ val(meta), path(fasta) ]
        ch_mtshift_bwaindex      // channel: [mandatory] [ val(meta), path(index) ]
        ch_mtshift_bwamem2index  // channel: [mandatory] [ val(meta), path(index) ]
        ch_mtshift_dictionary    // channel: [mandatory] [ val(meta), path(dict) ]
        ch_mtshift_fai           // channel: [mandatory] [ val(meta), path(fai) ]
        ch_mtshift_fasta         // channel: [mandatory] [ val(meta), path(fasta) ]
        val_mbuffer_mem          // integer: [mandatory] memory in megabytes
        val_platform             // string:  [mandatory] illumina or a different technology
        val_sort_threads         // integer: [mandatory] number of sorting threads

    main:
        ch_bwamem2_bam        = Channel.empty()
        ch_bwamem2_bai        = Channel.empty()
        ch_fastp_json         = Channel.empty()
        ch_markdup_metrics    = Channel.empty()
        ch_mt_bam_bai         = Channel.empty()
        ch_mt_marked_bam      = Channel.empty()
        ch_mt_marked_bai      = Channel.empty()
        ch_mtshift_bam_bai    = Channel.empty()
        ch_mtshift_marked_bam = Channel.empty()
        ch_mtshift_marked_bai = Channel.empty()
        ch_sentieon_bam       = Channel.empty()
        ch_sentieon_bai       = Channel.empty()
        ch_versions           = Channel.empty()

        if (!(params.skip_tools && params.skip_tools.split(',').contains('fastp'))) {
            FASTP (ch_reads, [], false, false, false)
            ch_reads = FASTP.out.reads
            ch_versions = ch_versions.mix(FASTP.out.versions)
            ch_fastp_json = FASTP.out.json
        }

        //
        // If input is bam
        //
        ch_alignments.map { meta, files ->
                    def new_id   = meta.sample
                    def new_meta = meta + [id:new_id, read_group:"\'@RG\\tID:" + new_id + "\\tPL:" + val_platform + "\\tSM:" + new_id + "\'"] - meta.subMap('lane')
                    return [new_meta, files].flatten()
                }
                .map { it -> [it[0], it[1]] }
                .set{ch_input_bam}

        ch_alignments.map { meta, files ->
                    def new_id   = meta.sample
                    def new_meta = meta + [id:new_id, read_group:"\'@RG\\tID:" + new_id + "\\tPL:" + val_platform + "\\tSM:" + new_id + "\'"] - meta.subMap('lane')
                    return [new_meta, files].flatten()
                }
                .map { it -> [it[0], it[2]] }
                .set{ch_input_bai}

        if (params.aligner.matches("bwamem2|bwa|bwameme")) {
            ALIGN_BWA_BWAMEM2_BWAMEME (             // Triggered when params.aligner is set as bwamem2 or bwa or bwameme
                ch_reads,
                ch_genome_bwaindex,
                ch_genome_bwamem2index,
                ch_genome_bwamemeindex,
                ch_genome_fasta,
                ch_genome_fai,
                val_mbuffer_mem,
                val_platform,
                val_sort_threads
            )
            ch_bwamem2_bam     = ALIGN_BWA_BWAMEM2_BWAMEME.out.marked_bam
            ch_bwamem2_bai     = ALIGN_BWA_BWAMEM2_BWAMEME.out.marked_bai
            ch_markdup_metrics = ALIGN_BWA_BWAMEM2_BWAMEME.out.metrics
            ch_versions        = ch_versions.mix(ALIGN_BWA_BWAMEM2_BWAMEME.out.versions)
        } else if (params.aligner.equals("sentieon")) {
            ALIGN_SENTIEON (                        // Triggered when params.aligner is set as sentieon
                ch_reads,
                ch_genome_fasta,
                ch_genome_fai,
                ch_genome_bwaindex,
                val_platform
            )
            ch_sentieon_bam    = ALIGN_SENTIEON.out.marked_bam
            ch_sentieon_bai    = ALIGN_SENTIEON.out.marked_bai
            ch_versions     = ch_versions.mix(ALIGN_SENTIEON.out.versions)
        }

        ch_genome_marked_bam = Channel.empty().mix(ch_bwamem2_bam, ch_sentieon_bam, ch_input_bam)
        ch_genome_marked_bai = Channel.empty().mix(ch_bwamem2_bai, ch_sentieon_bai, ch_input_bai)
        ch_genome_bam_bai    = ch_genome_marked_bam.join(ch_genome_marked_bai, failOnMismatch:true, failOnDuplicate:true)

        // PREPARING READS FOR MT ALIGNMENT

        if (params.analysis_type.matches("wgs|mito") || params.run_mt_for_wes) {
            CONVERT_MT_BAM_TO_FASTQ (
                ch_genome_bam_bai,
                ch_genome_fasta,
                ch_genome_fai,
                ch_genome_dictionary
            )

            ALIGN_MT (
                CONVERT_MT_BAM_TO_FASTQ.out.fastq,
                CONVERT_MT_BAM_TO_FASTQ.out.bam,
                ch_mt_bwaindex,
                ch_mt_bwamem2index,
                ch_mt_fasta,
                ch_mt_dictionary,
                ch_mt_fai
            )

            ALIGN_MT_SHIFT (
                CONVERT_MT_BAM_TO_FASTQ.out.fastq,
                CONVERT_MT_BAM_TO_FASTQ.out.bam,
                ch_mtshift_bwaindex,
                ch_mtshift_bwamem2index,
                ch_mtshift_fasta,
                ch_mtshift_dictionary,
                ch_mtshift_fai
            )

            ch_mt_marked_bam      = ALIGN_MT.out.marked_bam
            ch_mt_marked_bai      = ALIGN_MT.out.marked_bai
            ch_mt_bam_bai         = ch_mt_marked_bam.join(ch_mt_marked_bai, failOnMismatch:true, failOnDuplicate:true)
            ch_mtshift_marked_bam = ALIGN_MT_SHIFT.out.marked_bam
            ch_mtshift_marked_bai = ALIGN_MT_SHIFT.out.marked_bai
            ch_mtshift_bam_bai    = ch_mtshift_marked_bam.join(ch_mtshift_marked_bai, failOnMismatch:true, failOnDuplicate:true)
            ch_versions           = ch_versions.mix(ALIGN_MT.out.versions,
                                        ALIGN_MT_SHIFT.out.versions,
                                        CONVERT_MT_BAM_TO_FASTQ.out.versions)
        }

        if (params.save_mapped_as_cram) {
            SAMTOOLS_VIEW( ch_genome_bam_bai, ch_genome_fasta, [] )
            ch_versions   = ch_versions.mix(SAMTOOLS_VIEW.out.versions)
        }
    emit:
        fastp_json         = ch_fastp_json         // channel: [ val(meta), path(json) ]
        genome_marked_bam  = ch_genome_marked_bam  // channel: [ val(meta), path(bam) ]
        genome_marked_bai  = ch_genome_marked_bai  // channel: [ val(meta), path(bai) ]
        genome_bam_bai     = ch_genome_bam_bai     // channel: [ val(meta), path(bam), path(bai) ]
        markdup_metrics    = ch_markdup_metrics    // channel: [ val(meta), path(txt) ]
        mt_marked_bam      = ch_mt_marked_bam      // channel: [ val(meta), path(bam) ]
        mt_marked_bai      = ch_mt_marked_bai      // channel: [ val(meta), path(bai) ]
        mt_bam_bai         = ch_mt_bam_bai         // channel: [ val(meta), path(bam), path(bai) ]
        mtshift_marked_bam = ch_mtshift_marked_bam // channel: [ val(meta), path(bam) ]
        mtshift_marked_bai = ch_mtshift_marked_bai // channel: [ val(meta), path(bai) ]
        mtshift_bam_bai    = ch_mtshift_bam_bai    // channel: [ val(meta), path(bam), path(bai) ]
        versions           = ch_versions           // channel: [ path(versions.yml) ]
}
