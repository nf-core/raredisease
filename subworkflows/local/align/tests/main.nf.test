nextflow_workflow {

    name "Test Workflow ALIGN"
    script "subworkflows/local/align/main.nf"
    workflow "ALIGN"
    tag "subworkflows"
    tag "align"
    config "./nextflow.config"

    setup {
        run("BWAMEM2_INDEX", alias: 'MT_MEM2') {
            script "modules/nf-core/bwamem2/index/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [id:'mt'],
                    file('https://raw.githubusercontent.com/nf-core/test-datasets/raredisease/reference/reference_mt.fa', checkIfExists: true)
                ])
                """
            }
        }
        run("BWAMEM2_INDEX", alias: 'SHIFTMT_MEM2') {
            script "modules/nf-core/bwamem2/index/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [id:'shiftmt'],
                    file('https://raw.githubusercontent.com/nf-core/test-datasets/raredisease/reference/reference_shift.fasta', checkIfExists: true)
                ])
                """
            }
        }
    }

    test("align bwamem2") {

        setup {
            run("BWAMEM2_INDEX", alias: 'GENOME_MEM2') {
                script "modules/nf-core/bwamem2/index/main.nf"
                process {
                    """
                    input[0] = Channel.of([
                        [id:'genome'],
                        file('https://raw.githubusercontent.com/nf-core/test-datasets/raredisease/reference/reference.fasta', checkIfExists: true)
                    ])
                    """
                }
            }
        }

        when {
            params {
                aligner             = 'bwamem2'
                mt_aligner          = 'bwamem2'
                raredisease_testdata_base_path = 'https://raw.githubusercontent.com/nf-core/test-datasets/raredisease/'
                save_mapped_as_cram = true
                analysis_type       = "wgs"
                skip_tools          = null
                platform            = "illumina"
                min_trimmed_length  = 50
                mito_name           = "MT"
                outdir              = "$outputDir"
            }
            workflow {
                """
                input[0] = Channel.fromList([
                    [[ id:'earlycasualcaiman', sample:'earlycasualcaiman', single_end:false, num_lanes:1, read_group: "'@RG\\\\tID:earlycasualcaiman\\\\tPL:illumina\\\\tSM:earlycasualcaiman'", 'lane':1, 'sex':1, 'phenotype':1, 'paternal':0, 'maternal':0, 'case_id':'justhusky'  ], // meta map
                    [
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_earlycasualcaiman_XXXXXX_1.fastq.gz', checkIfExists: true),
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_earlycasualcaiman_XXXXXX_2.fastq.gz', checkIfExists: true)
                    ]],
                    [[ id:'hugelymodelbat', sample:'hugelymodelbat', single_end:false, num_lanes:1, read_group: "'@RG\\\\tID:hugelymodelbat\\\\tPL:illumina\\\\tSM:hugelymodelbat'", 'lane':1, 'sex':1, 'phenotype':2, 'paternal':'earlycasualcaiman', 'maternal':'slowlycivilbuck', 'case_id':'justhusky'  ], // meta map
                    [
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_hugelymodelbat_XXXXXX_1.fastq.gz', checkIfExists: true),
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_hugelymodelbat_XXXXXX_2.fastq.gz', checkIfExists: true)
                    ]],
                    [[ id:'slowlycivilbuck', sample:'slowlycivilbuck', single_end:false, num_lanes:1, read_group: "'@RG\\\\tID:slowlycivilbuck\\\\tPL:illumina\\\\tSM:slowlycivilbuck'",  'lane':1, 'sex':2, 'phenotype':1, 'paternal':0, 'maternal':0, 'case_id':'justhusky'  ], // meta map
                    [
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_slowlycivilbuck_XXXXXX_1.fastq.gz', checkIfExists: true),
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_slowlycivilbuck_XXXXXX_2.fastq.gz', checkIfExists: true)
                    ]]
                ])
                input[1]  = Channel.empty()
                input[2]  = Channel.of([[id:'genome'], [file(params.raredisease_testdata_base_path+ 'reference/reference.fasta', checkIfExists: true)]])
                input[3]  = Channel.of([[id:'genome'], [file(params.raredisease_testdata_base_path+ 'reference/reference.fasta.fai', checkIfExists: true)]])
                input[4]  = [[:],[]]
                input[5]  = GENOME_MEM2.out.index
                input[6]  = [[:],[]]
                input[7]  = Channel.of([[id:'genome'], [file(params.raredisease_testdata_base_path+ 'reference/reference.dict', checkIfExists: true)]])
                input[8]  = [[:],[]]
                input[9]  = MT_MEM2.out.index
                input[10] = Channel.of([[id:'mt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.dict', checkIfExists: true)]])
                input[11] = Channel.of([[id:'mt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.fa.fai', checkIfExists: true)]])
                input[12] = Channel.of([[id:'mt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.fa', checkIfExists: true)]])
                input[13] = [[:],[]]
                input[14] = SHIFTMT_MEM2.out.index
                input[15] = Channel.of([[id:'shiftmt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.dict', checkIfExists: true)]])
                input[16] = Channel.of([[id:'shiftmt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.fa.fai', checkIfExists: true)]])
                input[17] = Channel.of([[id:'shiftmt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.fa', checkIfExists: true)]])
                input[18] = 3072
                input[19] = "illumina"
                input[20] = 4
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                        workflow.out.genome_bam_bai.collect { meta, bamfile, index -> [ meta, bam(bamfile).getHeaderMD5(), index ] },
                        workflow.out.mt_bam_bai.collect { meta, bamfile, index -> [ meta, bam(bamfile).getReadsMD5(), index ] },
                        workflow.out.mtshift_bam_bai.collect { meta, bamfile, index -> [ meta, bam(bamfile).getReadsMD5(), index ] },
                        workflow.out.versions
                    ).match()
                }
            )
        }
    }

    test("align bwameme") {

        setup {
            run("BWAMEME_INDEX", alias: 'GENOME_MEME') {
                script "modules/nf-core/bwameme/index/main.nf"
                process {
                    """
                    input[0] = Channel.of([
                        [id:'genome'],
                        file('https://raw.githubusercontent.com/nf-core/test-datasets/raredisease/reference/reference.fasta', checkIfExists: true)
                    ])
                    """
                }
            }
        }

        when {
            params {
                aligner             = 'bwameme'
                mt_aligner          = 'bwamem2'
                raredisease_testdata_base_path = 'https://raw.githubusercontent.com/nf-core/test-datasets/raredisease/'
                save_mapped_as_cram = true
                analysis_type       = "wgs"
                skip_tools          = null
                platform            = "illumina"
                min_trimmed_length  = 50
                mito_name           = "MT"
                outdir              = "$outputDir"
            }
            workflow {
                """
                input[0] = Channel.fromList([
                    [[ id:'earlycasualcaiman', sample:'earlycasualcaiman', single_end:false, num_lanes:1, read_group: "'@RG\\\\tID:earlycasualcaiman\\\\tPL:illumina\\\\tSM:earlycasualcaiman'", 'lane':1, 'sex':1, 'phenotype':1, 'paternal':0, 'maternal':0, 'case_id':'justhusky'  ], // meta map
                    [
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_earlycasualcaiman_XXXXXX_1.fastq.gz', checkIfExists: true),
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_earlycasualcaiman_XXXXXX_2.fastq.gz', checkIfExists: true)
                    ]],
                    [[ id:'hugelymodelbat', sample:'hugelymodelbat', single_end:false, num_lanes:1, read_group: "'@RG\\\\tID:hugelymodelbat\\\\tPL:illumina\\\\tSM:hugelymodelbat'", 'lane':1, 'sex':1, 'phenotype':2, 'paternal':'earlycasualcaiman', 'maternal':'slowlycivilbuck', 'case_id':'justhusky'  ], // meta map
                    [
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_hugelymodelbat_XXXXXX_1.fastq.gz', checkIfExists: true),
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_hugelymodelbat_XXXXXX_2.fastq.gz', checkIfExists: true)
                    ]],
                    [[ id:'slowlycivilbuck', sample:'slowlycivilbuck', single_end:false, num_lanes:1, read_group: "'@RG\\\\tID:slowlycivilbuck\\\\tPL:illumina\\\\tSM:slowlycivilbuck'",  'lane':1, 'sex':2, 'phenotype':1, 'paternal':0, 'maternal':0, 'case_id':'justhusky'  ], // meta map
                    [
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_slowlycivilbuck_XXXXXX_1.fastq.gz', checkIfExists: true),
                        file(params.raredisease_testdata_base_path+ '/testdata/1_171015_HHT5NDSXX_slowlycivilbuck_XXXXXX_2.fastq.gz', checkIfExists: true)
                    ]]
                ])
                input[1]  = Channel.empty()
                input[2]  = Channel.of([[id:'genome'], [file(params.raredisease_testdata_base_path+ 'reference/reference.fasta', checkIfExists: true)]])
                input[3]  = Channel.of([[id:'genome'], [file(params.raredisease_testdata_base_path+ 'reference/reference.fasta.fai', checkIfExists: true)]])
                input[4]  = [[:],[]]
                input[5]  = [[:],[]]
                input[6]  = GENOME_MEME.out.index
                input[7]  = Channel.of([[id:'genome'], [file(params.raredisease_testdata_base_path+ 'reference/reference.dict', checkIfExists: true)]])
                input[8]  = [[:],[]]
                input[9]  = MT_MEM2.out.index
                input[10] = Channel.of([[id:'mt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.dict', checkIfExists: true)]])
                input[11] = Channel.of([[id:'mt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.fa.fai', checkIfExists: true)]])
                input[12] = Channel.of([[id:'mt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.fa', checkIfExists: true)]])
                input[13] = [[:],[]]
                input[14] = SHIFTMT_MEM2.out.index
                input[15] = Channel.of([[id:'shiftmt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.dict', checkIfExists: true)]])
                input[16] = Channel.of([[id:'shiftmt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.fa.fai', checkIfExists: true)]])
                input[17] = Channel.of([[id:'shiftmt'], [file(params.raredisease_testdata_base_path+ 'reference/reference_mt.fa', checkIfExists: true)]])
                input[18] = 3072
                input[19] = "illumina"
                input[20] = 4
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                        workflow.out.genome_bam_bai.collect { meta, bamfile, index -> [ meta, bam(bamfile).getHeaderMD5(), index ] },
                        workflow.out.mt_bam_bai.collect { meta, bamfile, index -> [ meta, bam(bamfile).getReadsMD5(), index ] },
                        workflow.out.mtshift_bam_bai.collect { meta, bamfile, index -> [ meta, bam(bamfile).getReadsMD5(), index ] },
                        workflow.out.versions
                    ).match()
                }
            )
        }
    }

}
