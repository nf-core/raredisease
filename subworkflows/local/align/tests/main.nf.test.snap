{
    "align bwamem2": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "slowlycivilbuck",
                            "sample": "slowlycivilbuck",
                            "single_end": false,
                            "num_lanes": 1,
                            "read_group": "'@RG\\tID:slowlycivilbuck\\tPL:illumina\\tSM:slowlycivilbuck'",
                            "sex": 2,
                            "phenotype": 1,
                            "paternal": 0,
                            "maternal": 0,
                            "case_id": "justhusky"
                        }
                    },
                    "8aa7930105c6aa55f96e3719de968144",
                    "slowlycivilbuck_sorted_md.bam.bai:md5,b6c5d2736032f5e1cb190853ca366afe"
                ]
            ],
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "slowlycivilbuck",
                            "sample": "slowlycivilbuck",
                            "single_end": false,
                            "num_lanes": 1,
                            "read_group": "'@RG\\tID:slowlycivilbuck\\tPL:illumina\\tSM:slowlycivilbuck'",
                            "sex": 2,
                            "phenotype": 1,
                            "paternal": 0,
                            "maternal": 0,
                            "case_id": "justhusky"
                        }
                    },
                    "a5d6892191a2c8c1f29966448623ca1b",
                    "slowlycivilbuck_sorted_merged_md_sorted.bam.bai:md5,35373ec0e388ae91025a4eec9897434e"
                ]
            ],
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "slowlycivilbuck",
                            "sample": "slowlycivilbuck",
                            "single_end": false,
                            "num_lanes": 1,
                            "read_group": "'@RG\\tID:slowlycivilbuck\\tPL:illumina\\tSM:slowlycivilbuck'",
                            "sex": 2,
                            "phenotype": 1,
                            "paternal": 0,
                            "maternal": 0,
                            "case_id": "justhusky"
                        }
                    },
                    "427b5e476d84c5277d2d24210800ad60",
                    "slowlycivilbuck_sorted_merged_md_sorted_shifted.bam.bai:md5,6c663c5661749b25782ac78a3ba1f242"
                ]
            ],
            [
                "versions.yml:md5,04cb590a101237e8d8726039ee466819",
                "versions.yml:md5,210790d64830e81c512ff898e8d888db",
                "versions.yml:md5,3524532e0d268745c6d6c785bc7e734e",
                "versions.yml:md5,3cdf4ec614502fb38c2475c5475c3e0c",
                "versions.yml:md5,3efddfe43fdf9bf446e23d1526a3973a",
                "versions.yml:md5,6cbf152be76b9a14bb7eccb2658cabb3",
                "versions.yml:md5,6f9a56c7077343a78b6b2a9bb555631b",
                "versions.yml:md5,753dc9d8832a5f5a1e10237d78e157f2",
                "versions.yml:md5,7b799d18a80d6d144122cc97f83748c3",
                "versions.yml:md5,8191c0426b9e36d914ea3f63bf2428e9",
                "versions.yml:md5,81cee5ff00c008209a5c6fb06ac57de5",
                "versions.yml:md5,81cee5ff00c008209a5c6fb06ac57de5",
                "versions.yml:md5,81cee5ff00c008209a5c6fb06ac57de5",
                "versions.yml:md5,87e7da5e82b70b42f06b30a245a5750e",
                "versions.yml:md5,914929705542ec69faf26ad6c672e2af",
                "versions.yml:md5,9c218b25dd02dc279b1ce6fad9ae182b",
                "versions.yml:md5,b1c46403177c797d5fc304eebf16731c",
                "versions.yml:md5,b511e8fd8c8b46c5470d3e3e79673502",
                "versions.yml:md5,b9749a7d836c638ad5deb0f8b208c7ec",
                "versions.yml:md5,c31eb465488334212645bd43296d8bd7",
                "versions.yml:md5,d19b223e2ea8c907b23c9b301c15f3fb",
                "versions.yml:md5,e092f50493ce9a5725381d53e24519d0",
                "versions.yml:md5,e8fa9e5b444276ecb6b7af4cb7a86247",
                "versions.yml:md5,f8dd98ba8455bb8d00379fba39cc9e65"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-05-23T15:53:23.463213506"
    },
    "align bwameme": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "slowlycivilbuck",
                            "sample": "slowlycivilbuck",
                            "single_end": false,
                            "num_lanes": 1,
                            "read_group": "'@RG\\tID:slowlycivilbuck\\tPL:illumina\\tSM:slowlycivilbuck'",
                            "sex": 2,
                            "phenotype": 1,
                            "paternal": 0,
                            "maternal": 0,
                            "case_id": "justhusky"
                        }
                    },
                    "b75a16ff1082ce42e9a9ac046b509181",
                    "slowlycivilbuck_sorted_md.bam.bai:md5,ff8eb902e42e6aeec775b9b412252ab6"
                ]
            ],
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "slowlycivilbuck",
                            "sample": "slowlycivilbuck",
                            "single_end": false,
                            "num_lanes": 1,
                            "read_group": "'@RG\\tID:slowlycivilbuck\\tPL:illumina\\tSM:slowlycivilbuck'",
                            "sex": 2,
                            "phenotype": 1,
                            "paternal": 0,
                            "maternal": 0,
                            "case_id": "justhusky"
                        }
                    },
                    "a5d6892191a2c8c1f29966448623ca1b",
                    "slowlycivilbuck_sorted_merged_md_sorted.bam.bai:md5,35373ec0e388ae91025a4eec9897434e"
                ]
            ],
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "slowlycivilbuck",
                            "sample": "slowlycivilbuck",
                            "single_end": false,
                            "num_lanes": 1,
                            "read_group": "'@RG\\tID:slowlycivilbuck\\tPL:illumina\\tSM:slowlycivilbuck'",
                            "sex": 2,
                            "phenotype": 1,
                            "paternal": 0,
                            "maternal": 0,
                            "case_id": "justhusky"
                        }
                    },
                    "427b5e476d84c5277d2d24210800ad60",
                    "slowlycivilbuck_sorted_merged_md_sorted_shifted.bam.bai:md5,6c663c5661749b25782ac78a3ba1f242"
                ]
            ],
            [
                "versions.yml:md5,04cb590a101237e8d8726039ee466819",
                "versions.yml:md5,210790d64830e81c512ff898e8d888db",
                "versions.yml:md5,3524532e0d268745c6d6c785bc7e734e",
                "versions.yml:md5,3cdf4ec614502fb38c2475c5475c3e0c",
                "versions.yml:md5,3efddfe43fdf9bf446e23d1526a3973a",
                "versions.yml:md5,6cbf152be76b9a14bb7eccb2658cabb3",
                "versions.yml:md5,6f9a56c7077343a78b6b2a9bb555631b",
                "versions.yml:md5,753dc9d8832a5f5a1e10237d78e157f2",
                "versions.yml:md5,7b799d18a80d6d144122cc97f83748c3",
                "versions.yml:md5,81cee5ff00c008209a5c6fb06ac57de5",
                "versions.yml:md5,81cee5ff00c008209a5c6fb06ac57de5",
                "versions.yml:md5,81cee5ff00c008209a5c6fb06ac57de5",
                "versions.yml:md5,87e7da5e82b70b42f06b30a245a5750e",
                "versions.yml:md5,914929705542ec69faf26ad6c672e2af",
                "versions.yml:md5,9c218b25dd02dc279b1ce6fad9ae182b",
                "versions.yml:md5,aef5efcb8852a854022d16b939b52d17",
                "versions.yml:md5,b1c46403177c797d5fc304eebf16731c",
                "versions.yml:md5,b511e8fd8c8b46c5470d3e3e79673502",
                "versions.yml:md5,b9749a7d836c638ad5deb0f8b208c7ec",
                "versions.yml:md5,c31eb465488334212645bd43296d8bd7",
                "versions.yml:md5,d19b223e2ea8c907b23c9b301c15f3fb",
                "versions.yml:md5,e092f50493ce9a5725381d53e24519d0",
                "versions.yml:md5,e8fa9e5b444276ecb6b7af4cb7a86247",
                "versions.yml:md5,f8dd98ba8455bb8d00379fba39cc9e65"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-05-23T15:59:18.726434076"
    }
}